### PR TITLE
mfem: fix concretization

### LINKED ISSUE #7916: laghos cannot be build
```
$ spack install laghos
==> Error: An unsatisfiable version constraint has been detected for spec:

    hypre@2.14.0%gcc@7.2.0~int64+internal-superlu+mpi+shared arch=linux-gentoo2-x86_64


while trying to concretize the partial spec:

    mfem@3.3.2%gcc@7.2.0~conduit~debug~examples~gnutls+gzstream+hypre~lapack~libunwind+metis~miniapps~mpfr+mpi~netcdf~openmp~petsc~shared+static~suite-sparse~sundials~superlu-dist~threadsafe timer=auto arch=linux-gentoo2-x86_64
        ^openmpi@3.0.1%gcc@7.2.0~cuda fabrics= ~java~memchecker~pmi schedulers= ~sqlite3~thread_multiple+vt arch=linux-gentoo2-x86_64
            ^hwloc@:1.999
                ^pkgconfig
            ^numactl
                ^autoconf
                    ^m4@1.4.6:
                    ^perl
                        ^gdbm
                            ^readline
                                ^ncurses
                ^automake
                ^libtool
            ^zlib


mfem requires hypre version 2.10.0:2.13.999, but spec asked for 2.14.0
```

Related to #7903

### PR DESCRIPTION
**Pull request by junghans**
Fix #7916
Fix proxyapps/proxy-ci#13

### CODE REVIEW

File: var/spack/repos/builtin/packages/mfem/package.py [hunk 1]
```diff
@@ -163,8 +163,8 @@ class Mfem(Package):
     conflicts('timer=mpi', when='~mpi')
 
     depends_on('mpi', when='+mpi')
-    depends_on('hypre', when='+mpi')
-    depends_on('hypre@2.10.0:2.13.999', when='@:3.3.2')
+    depends_on('hypre@2.10.0:2.13.999', when='@:3.3.2+mpi')
+    depends_on('hypre@2.14:', when='@3.3.3:+mpi')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/mfem/package.py]: Is it true that `mfem@3.3.3` can't be built with `hypre@2.13`? If concretization is the only problem, the first line of your change should be sufficient.

Comment (tzanio) [reply var/spack/repos/builtin/packages/mfem/package.py]: 3.3.3 is the internal number we use for development after 3.3.2, which currently is the same as `mfem@develop`. There will be no official release with that number, the next release will be 3.4.

(If you are curious, versions with even patch number correspond to official release, while odd patch numbers correspond to development)

So:
- `mfem@3.3.2` can be built with `hypre@2.13` but not `hypre@2.14`
- `mfem@develop` and `mfem@3.4` (when it comes out soon) can be built with any hypre version, including `hypre@2.14`

**Comment by junghans** [comment]
@adamjstewart, is that ok now?

**Comment by adamjstewart** [comment]
So no version can be built with `hypre` < 2.10.0?

**Comment by junghans** [comment]
That lower bound was introduced in #7903, but it doesn’t seem necessary!

**Comment by junghans** [comment]
@adamjstewart done

**Pull request merged**